File: CCC '07/Senior/S3 - Friends.py
# https://dmoj.ca/problem/ccc07s3
# CCC '07 S3 - Friends

from collections import deque

n = int(input())

def theory(start, end, adj):
    dq = deque()
    dq.append((start, 0))

    visited = {}

    for i in adj.values():
        for j in i:
            if j not in visited:
                visited[j] = False

    while len(dq) > 0:
        curNode, dis = dq.pop()
        if curNode == end:
            return f"Yes {dis - 1}"
        
        if visited[curNode]:
            continue

        visited[curNode] = True

        for i in adj[curNode]:
            if not visited[i]:
                dq.append((i, dis+1))

    return f"No"


adj = {}

for i in range(n):
    a, b = map(int, input().split())
    if a in adj:
        adj[a].append(b)
    else:
        adj[a] = [b]

while True:
    start, end = map(int, input().split())

    if start == 0 and end == 0:
        break

    print(theory(start, end, adj))

File: CCC '07/Senior/S3 - Friends.cpp
// https://dmoj.ca/problem/ccc07s3
// CCC '07 S3 - Friends

#include <iostream>
using namespace std;

const int maxn = 1e4+5;

int ppl[maxn];
bool vis[maxn];

int n,tmp1,tmp2;

int main() {
	cin>>n;
	for(int i = 0; i < n; i++){
		cin>>tmp1>>tmp2;
		ppl[tmp1] = tmp2;
	}
	cin>>tmp1>>tmp2;
	while(!(tmp1 == 0 && tmp2==0)){
		int x = tmp1; int y =  tmp2; int dis = -1;
		
		while(x!=y && !vis[x]){
			vis[x] = true;
			x = ppl[x];
			dis++;
		}
		
		if(x==y){
			cout<<"Yes "<<dis<<endl;
		}
		else{
			cout<<"No"<<endl;
		}
		for(int i = 0; i<=n; i++){
			vis[i] = false;
		}
		cin>>tmp1>>tmp2;
	}
	return 0;
}